### PR TITLE
Add default processing

### PR DESCRIPTION
**Pull request by antgonza**
First merge https://github.com/qiita-spots/qiita/pull/3110; then I'll update this PR.

This PR adds a single button to generate/populate based on our default pipelines, for example:
add-default-workflow

### CODE REVIEW

File: qiita_db/metadata_template/prep_template.py [hunk 1]
```diff
@@ -713,3 +713,192 @@ def modification_timestamp(self):
...
+                pred = _get_predecessors(pnode)
+                cxns = {x[0]: x[2]
+                        for x in wk.graph.get_edge_data(
+                            pnode, node)['connections'].connections}
+                data = [pnode, node, cxns]
+                if pred is None:
+                    pred = [data]
+                else:
+                    pred.append(data)
+                return pred
```
Comment (wasade) [inline qiita_db/metadata_template/prep_template.py:776]: I *think* that having the `return` in scope of the `for` would imply a node can only have a single predecessor. Is that correct for both interpretation and intent?

File: qiita_db/metadata_template/test/test_prep_template.py [hunk 2]
```diff
@@ -1340,11 +1340,26 @@ def test_artifact_setter_error(self):
...
     def test_artifact_setter(self):
         pt = qdb.metadata_template.prep_template.PrepTemplate.create(
-            self.metadata, self.test_study, self.data_type_id)
+            self.metadata, self.test_study, '16S')
         self.assertEqual(pt.artifact, None)
         artifact = qdb.artifact.Artifact.create(
             self.filepaths, "FASTQ", prep_template=pt)
         self.assertEqual(pt.artifact, artifact)
+
+        # here we can test that we can properly create a workflow
```
Comment (wasade) [inline qiita_db/metadata_template/test/test_prep_template.py:1349]: This seems like a limited test given the complexity of the function, are there edge cases beyond the `ValueError` check below that should be examined?

File: qiita_db/metadata_template/prep_template.py [hunk 3]
```diff
@@ -713,3 +713,192 @@ def modification_timestamp(self):
...
+
+        Raises
+        ------
+        ValueError
+            If this preparation doesn't have valid workflows
+            If this preparation has been fully processed
+        """
+        # helper functions to avoid duplication of code
+
+        def _get_node_info(node):
```
[the remark this answers is not in the document]
Comment (antgonza) [reply qiita_db/metadata_template/prep_template.py:738]: I could add but when this is called, wk is "global" within this method; do you think is better to pass?

File: qiita_db/metadata_template/prep_template.py [hunk 4]
```diff
@@ -713,3 +713,192 @@ def modification_timestamp(self):
...
+                parent_cmd_name = pcmd.name
+                parent_merging_scheme = pcmd.merging_scheme
+
+            return qdb.util.human_merging_scheme(
+                ccmd.name, ccmd.merging_scheme, parent_cmd_name,
+                parent_merging_scheme, cparams, [], pparams)
+
+        def _get_predecessors(node):
+            # recursive method to get predecessors of a given node
+            for pnode in wk.graph.predecessors(node):
```
[the remark this answers is not in the document]
Comment (antgonza) [reply qiita_db/metadata_template/prep_template.py:766]: Yup, same, they exists in `add_default_workflow` and the functions are defined and called within that scope so they are "global" ... in fact, we might not even need to pass node. Anyway, I'm OK changing to whatever you think is best, just let me know.

File: qiita_db/metadata_template/prep_template.py (hunk 1 above)
Comment (antgonza) [reply qiita_db/metadata_template/prep_template.py:776]: Yes!! Note that this is a recursive function (line 767 calls itself) as we need the "grandparent" information to create the "parent" information.

File: qiita_db/metadata_template/prep_template.py [hunk 5]
```diff
@@ -713,3 +713,192 @@ def modification_timestamp(self):
...
+            for node, degree in wk.graph.out_degree():
+                if degree != 0:
+                    continue
+                mscheme = _get_node_info(node)
+                if mscheme not in merging_schemes:
+                    missing_artifacts[wk][mscheme] = node
+            if not missing_artifacts[wk]:
+                del missing_artifacts[wk]
+        if not missing_artifacts:
+            raise ValueError('This preparation is complete')
```
[the remark this answers is not in the document]
Comment (antgonza) [reply qiita_db/metadata_template/prep_template.py:823]: You are right, I'll add it.

File: qiita_db/metadata_template/test/test_prep_template.py (hunk 2 above)
Comment (antgonza) [reply qiita_db/metadata_template/test/test_prep_template.py:1349]: Good point, I'll add more tests.

**Review by antgonza — COMMENTED** [review]
@wasade, thank you for detailed review, I have some questions; then, I'll issue fixes to your comments.

File: qiita_db/metadata_template/prep_template.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wasade) [reply qiita_db/metadata_template/prep_template.py:738]: it may be nice to either pass both or pass neither as `node` is also non-local from what I can tell from the function call

File: qiita_db/metadata_template/prep_template.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wasade) [reply qiita_db/metadata_template/prep_template.py:766]: passing both may be nice to be explicit, tracing non-local variables can get confusing quickly

File: qiita_db/metadata_template/prep_template.py (hunk 1 above)
Comment (wasade) [reply qiita_db/metadata_template/prep_template.py:776]: Ya, but just so I understand, is it correct that a node only ever has a single parent?

File: qiita_db/metadata_template/prep_template.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wasade) [reply qiita_db/metadata_template/prep_template.py:823]: Thanks!

File: qiita_db/metadata_template/test/test_prep_template.py (hunk 2 above)
Comment (wasade) [reply qiita_db/metadata_template/test/test_prep_template.py:1349]: Thanks!

**Comment by mestaki** [comment]
Wasn't sure if this should go here or have its own separate issue, let me know and happy create a new issue.

As per the recent email thread on this topic: the Standard Workflow option is great, but without a brief description of what it actually is/does it might be confusing for users. Adding a brief description of this would be thus very useful. Happy to contribute to this. 
Option 1: Add a short description on top of the main network processing page, 
Option 2: A brief "hover-over" text box appears over the "Standard Workflow" button,
Option 3: A detailed description in the main help files and just provide a link to that in the network processing page?
Option 4: something else?

Depending on the options the description I can make the text brief or more detailed.

**Comment by antgonza** [comment]
Thank you @mestaki; I think 1 and 3 would be great. If you send me an email with that I can open a PR adding those texts so you can review there ... or you can open a PR with those changes and I can review; whatever works best. BTW no need to open a new issue.

File: qiita_db/metadata_template/test/test_prep_template.py (hunk 2 above)
Comment (antgonza) [reply qiita_db/metadata_template/test/test_prep_template.py:1349]: Just FYI, I checked everything that came to mind and I think those "simple" tests actually cover all scenarios by having to create a full pipeline and then checking that full pipeline and not having anything else to create.

File: qiita_db/metadata_template/prep_template.py (hunk 1 above)
Comment (antgonza) [reply qiita_db/metadata_template/prep_template.py:778]: Thank you for double checking this; better to be safe. 

Yes, I spent sometime looking at this and this is the expected behavior; basically, by being recursive it will go down to the origin of the workflow (always a single artifact) and return (`if pred is None:`) the single only parent, then it will start coming out of the recursion and appending in order the children jobs (`else:`). In case it helps, we tested this in qiita-rc and everything seem fine.

**Review by antgonza — COMMENTED** [review]
Thank you @wasade.

**Comment by wasade** [comment]
Good to merge?

**Comment by antgonza** [comment]
Yes, thank you. The tutorial/text updates will be done in a separate PR (together with the release notes).
